terraform/transform_resource.go

```diff
@@ -134,8 +134,34 @@ func (n *graphNodeExpandedResource) DependableName() []string {
 
 // GraphNodeDependent impl.
 func (n *graphNodeExpandedResource) DependentOn() []string {
-	config := &GraphNodeConfigResource{Resource: n.Resource}
-	return config.DependentOn()
+	configNode := &GraphNodeConfigResource{Resource: n.Resource}
+	result := configNode.DependentOn()
+
+	// Walk the variables to find any count-specific variables we depend on.
+	configNode.VarWalk(func(v config.InterpolatedVariable) {
+		rv, ok := v.(*config.ResourceVariable)
+		if !ok {
+			return
+		}
+
+		// We only want ourselves
+		if rv.ResourceId() != n.Resource.Id() {
+			return
+		}
+
+		// If this isn't a multi-access (which shouldn't be allowed but
+		// is verified elsewhere), then we depend on the specific count
+		// of this resource, ignoring ourself (which again should be
+		// validated elsewhere).
+		if rv.Index > -1 {
```

Comment:
Oh `Multi` means "referencing more than one" not "referencing _into_ a collection". In my proto-fix I had this as:

```go
			if rv, ok := v.(*config.ResourceVariable); ok {
				if rv.Multi && rv.Index >= 0 {
					// Skip self references
					if rv.ResourceId() == n.Resource.Id() && rv.Index == n.Index {
						continue
					}
					deps = append(deps, fmt.Sprintf("%s.%d", rv.ResourceId(), rv.Index))
				}
			}
```

Maybe tell me what concepts I had missing there, if any? Just for the learnings.